agg-ast/schema_derivation/src/schema_derivation_tests/match_expr.rs

```diff
@@ -2771,3 +2771,803 @@ mod string_ops {
+            keys: map! {
+                "foo".to_string() => Schema::AnyOf(set!(
+                    Schema::Array(Box::new(Schema::Atomic(Atomic::Integer))),
+                    Schema::Atomic(Atomic::Null)
+                )),
+                "bar".to_string() => NUMERIC_OR_NULL.clone(),
+            },
+            ..Default::default()
+        })),
+        input = r#"{"$match": {"$expr": {"$eq": [null, {"$arrayElemAt": ["$foo", "$bar"]}]}}}"#,
+        starting_schema = Schema::Document(Document {
+            keys: map! {
+                "foo".to_string() => Schema::AnyOf(set!(
+                    Schema::Atomic(Atomic::Null),
+                    Schema::Missing,
+                    Schema::Array(Box::new(Schema::Atomic(Atomic::Integer))),
+                )),
+                "bar".to_string() => Schema::Any
+            },
+            required: set!("foo".to_string(), "bar".to_string()),
+            ..Default::default()
+        })
+    );
+    test_derive_schema_for_match_stage!(
+        array_to_object_not_null,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "foo".to_string() => Schema::Array(Box::new(Schema::Array(Box::new(Schema::Any)))),
+            },
+            required: set!("foo".to_string()),
+            ..Default::default()
+        })),
+        input = r#"{"$match": {"$expr": {"$arrayToObject": "$foo"}}}"#,
+        ref_schema = Schema::Any
+    );
+    test_derive_schema_for_match_stage!(
+        array_to_object_maybe_null,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "foo".to_string() => Schema::AnyOf(set!(
+                    Schema::Array(Box::new(Schema::Array(Box::new(Schema::Any)))),
+                    Schema::Atomic(Atomic::Null),
+                ))
+            },
+            ..Default::default()
+        })),
+        input = r#"{"$match": {"$expr": {"$ne": [[], {"$arrayToObject": "$foo"}]}}}"#,
+        ref_schema = Schema::Any
+    );
+    test_derive_schema_for_match_stage!(
+        array_to_object_null,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "foo".to_string() => Schema::Atomic(Atomic::Null)
+            },
+            ..Default::default()
+        })),
+        input = r#"{"$match": {"$expr": {"$eq": [null, {"$objectToArray": "$foo"}]}}}"#,
+        ref_schema = Schema::Any
+    );
+    test_derive_schema_for_match_stage!(
+        merge_objects_single_arg_ref,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "foo".to_string() => Schema::AnyOf(set!(
+                    Schema::Array(Box::new(Schema::Document(Document::any()))),
+                    Schema::Document(Document::any())
+                ))
+            },
+            required: set!("foo".to_string()),
+            ..Default::default()
+        })),
+        input = r#"{"$match": {"$expr": {"$mergeObjects": "$foo"}}}"#,
```

Comment:
I am wholly surprised that this works with $foo being an array. I think most operators don't work that way. I love agg :joy:

Reply:
It is really bewildering. It works for an array or document, hence why neither can be assumed. Maybe it is usable in another operator that makes this make snese (like a windowing type deal)
```
derivation> db.aggregate([ { $documents: [ { _id: 1, "foo": { "a": 1 } }] }, { $project: { expr: { $mergeObjects: "$foo" } } }] )
[ { _id: 1, expr: { a: 1 } } ]
derivation> db.aggregate([ { $documents: [ { _id: 1, "foo": [{ "a": 1 }] }] }, { $project: { expr: { $mergeObjects: "$foo" } } }] )
[ { _id: 1, expr: { a: 1 } } ]
```